core/src/main/java/brooklyn/util/crypto/SecureKeys.java

```
@@ -19,182 +19,149 @@
-        return newKeyPairGenerator(algorithm, bits).generateKeyPair();
-    }
+    /** reads RSA or DSA / pem style private key files (viz {@link #toPem(KeyPair)}), extracting also the public key if possible
+     * @throws IllegalStateException on errors, in particular {@link PassphraseProblem} if that is the problem */
+    public static KeyPair readPem(InputStream input, final String passphrase) {
+        // TODO cache is only for fallback "reader" strategy 
+        byte[] cache = Streams.readFully(input);
+        input = new ByteArrayInputStream(cache);
 
-    /** returns a new keystore, of the default type, and initialized to be empty.
-     * (everyone always forgets to load(null,null).) */
-    public static KeyStore newKeyStore() {
         try {
-            KeyStore store = KeyStore.getInstance(KeyStore.getDefaultType());
-            store.load(null, null);
-            return store;
-        } catch (Exception e) {
-            throw Exceptions.propagate(e);
-        }
-    }
+            PEMParser pemParser = new PEMParser(new InputStreamReader(input));
+
+            Object object = pemParser.readObject();
+            pemParser.close();
+
+            JcaPEMKeyConverter converter = new JcaPEMKeyConverter().setProvider("BC");
+            KeyPair kp = null;
+            if (object==null) {
+                throw new IllegalStateException("PEM parsing failed: missing or invalid data");
+            } else if (object instanceof PEMEncryptedKeyPair) {
+                if (passphrase==null) throw new PassphraseProblem("passphrase required");
+                try {
+                    PEMDecryptorProvider decProv = new JcePEMDecryptorProviderBuilder().build(passphrase.toCharArray());
+                    kp = converter.getKeyPair(((PEMEncryptedKeyPair) object).decryptKeyPair(decProv));
+                } catch (Exception e) {
+                    Exceptions.propagateIfFatal(e);
+                    throw new PassphraseProblem("wrong passphrase");
```

I'd pass in the cause as well. Is there any more specific exception we can catch here? Feels risky that we could get other exceptions and swallow them, incorrectly reporting it as a "wrong passphrase".